### PR TITLE
Added SpecFunc::AccurateSum

### PR DESCRIPTION
**Pull request by regislebrun**
This method allows to compute the sum of a sequence of floats with an extended precision, using either MPFR or Kahan's method.

### CODE REVIEW

File: lib/src/Uncertainty/Model/DistributionImplementation.cxx [hunk 1]
```diff
@@ -1007,9 +1008,10 @@ Scalar DistributionImplementation::computeProbabilityGeneral(const Interval & in
         }
       } // j
       const Scalar cdf = computeCDF(c);
-      probability += (evenLower ? cdf : -cdf);
+      probabilities[i] = (evenLower ? cdf : -cdf);
     } // i
-
+    std::cerr << "probabilities=" << probabilities << std::endl;
```
Comment (jschueller) [inline lib/src/Uncertainty/Model/DistributionImplementation.cxx:1013]: leftover from debug?

File: python/test/t_SpecFunc_std.py [hunk 2]
```diff
@@ -38,3 +38,7 @@
     "ibeta(0.0, 1.0, 0.95, tail)=",
     ot.SpecFunc.RegularizedIncompleteBeta(0.0, 1.0, 0.95, True),
 )
+
+x = [1.0, 2.0**54, -2.0**54]
+print("sum(x)=", sum(x))
+print("accurate sum(x)=", ot.SpecFunc.AccurateSum(x))
```
Comment (josephmure) [inline python/test/t_SpecFunc_std.py:44]:
```suggestion
assert sum(x) == 0.0, "sum(x) nonzero"
assert ot.SpecFunc.AccurateSum(x) == 1.0, "accurate sum(x) not 1"
```

File: python/test/t_SpecFunc_std.expout [hunk 3]
```diff
@@ -60,3 +60,5 @@ ibeta(2.0, 0.0, 0.95)= 0.0
 ibeta(0.0, 1.0, 0.95)= 1.0
 ibeta(2.0, 0.0, 0.95, tail)= 1.0
 ibeta(0.0, 1.0, 0.95, tail)= 0.0
+sum(x)= 0.0
+accurate sum(x)= 1.0
```
Comment (josephmure) [inline python/test/t_SpecFunc_std.expout:64]:
```suggestion
```

**Review by josephmure — CHANGES_REQUESTED** [review]

File: lib/src/Uncertainty/Model/DistributionImplementation.cxx (hunk 1 above)
Comment (regislebrun) [reply lib/src/Uncertainty/Model/DistributionImplementation.cxx:1013]: Yep. Fixed.

File: lib/src/Uncertainty/Algorithm/Analytical/SystemFORM.cxx [hunk 4]
```diff
@@ -214,17 +217,15 @@ void SystemFORM::run()
...
-    const Scalar poincarreProbability = normal.computeCDF(-1.0 * regionBeta);
+    signedProbabilities[k] = sign * normal.computeCDF(-1.0 * regionBeta);
+    sign = -sign;
 
-    LOGINFO(OSS() << "SystemFORM: poincarre probability [" << k << "]=" << poincarreProbability);
-
-    // trick to update the poincarre sum
-    // we invert the sign of each consecutive term to follow the order of our recursive expansion of the Poincarre formula
-    eventProbability = -eventProbability + poincarreProbability;
+    LOGINFO(OSS() << "SystemFORM: poincarre probability [" << k << "]=" << signedProbabilities[k]);
```
[the remark this answers is not in the document]
Comment (regislebrun) [reply lib/src/Uncertainty/Algorithm/Analytical/SystemFORM.cxx:223]: @josephmure This one is from @jschueller and is fixed in another PR

**Review by josephmure — APPROVED** [review]

**Review by jschueller — APPROVED** [review]

**Comment by jschueller** [comment]
you can fix the conflicts now

**Comment by jschueller** [comment]
the $ms compiler is not happy :[

**Comment by regislebrun** [comment]
@jschueller I saw it. It must be linked with a stupid optimization simplifying Kahan's algorithm as if float arithmetic was associative and commutative. Maybe a pragma such as
```
#pragma optimize( "", off )
```
would do the job ([details](https://learn.microsoft.com/en-us/cpp/preprocessor/optimize?view=msvc-170)), as suggested [here](https://stackoverflow.com/questions/32718989/disable-optimization-on-specific-msvc-version)

**Comment by regislebrun** [comment]
@jschueller After inspection, it is not the fault of the MS compiler, the test was too stringent for Kahan's method. It was masked on Linux platform due to the use of the alternative MPFR implementation. Any hope to have it on Windows?

**Comment by josephmure** [comment]
@regislebrun If the test is too stringent, it should be relaxed IMO. Its purpose is to test that Kahan's method is well implemented so it should account for possible platform-related differences.

**Comment by regislebrun** [comment]
@josephmure You are right, and it is exactly what it does now. The Kahan method gives the exact same results on all the platforms and hopefully will continue to give the same results in the future ;-)

**Comment by jschueller** [comment]
on windows we distribute mingw binaries and mpfr is available there, so I'll probably wont bother testing mpfr with msvc

File: lib/src/Base/Func/SpecFunc/SpecFunc.cxx [hunk 5]
```diff
@@ -812,6 +812,13 @@ Scalar SpecFunc::Stirlerr(const UnsignedInteger n)
 }
 
 // Hypergeometric function of type (1,1): HyperGeom_1_1(p1, q1, x) = \sum_{n=0}^{\infty} [\prod_{k=0}^{n-1} (p1 + k) / (q1 + k)] * x^n / n!
+// To prevent Microsoft compiler for abusive optimization
+#ifdef _WIN32
+#ifdef _MSC_VER
+#pragma optimize( "", off )
```
Comment (jschueller) [inline lib/src/Base/Func/SpecFunc/SpecFunc.cxx:818]: so optimization can be re-enabled now ?
note that _MSC_VER implies _WIN32

**Review by jschueller — APPROVED** [review]

File: lib/src/Base/Func/SpecFunc/SpecFunc.cxx [hunk 6]
```diff
@@ -812,6 +812,7 @@ Scalar SpecFunc::Stirlerr(const UnsignedInteger n)
 }
 
 // Hypergeometric function of type (1,1): HyperGeom_1_1(p1, q1, x) = \sum_{n=0}^{\infty} [\prod_{k=0}^{n-1} (p1 + k) / (q1 + k)] * x^n / n!
+// To prevent Microsoft compiler for abusive optimization
```
Comment (jschueller) [inline lib/src/Base/Func/SpecFunc/SpecFunc.cxx:815]: drop this one

Comment (regislebrun) [reply lib/src/Base/Func/SpecFunc/SpecFunc.cxx]: done

**Pull request merged**